pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,640 @@
+
+    Returns:
+        callable: Reconstructed Fourier series with ``num_frequency`` frequencies.
+        This function is a purely classical function. Furthermore, it is fully
+        differentiable.
+    """
+    if not abs(int(num_frequency)) == num_frequency:
+        raise ValueError(f"num_frequency must be a non-negative integer, got {num_frequency}")
+
+    a = (num_frequency + 0.5) / np.pi
+    b = 0.5 / np.pi
+
+    shifts_pos = qml.math.arange(1, num_frequency + 1) / a
+    shifts_neg = -shifts_pos[::-1]
+    shifts = qml.math.concatenate([shifts_neg, [0.0], shifts_pos])
+    shifts = anp.asarray(shifts, like=interface)
+    f0 = fun(0.0) if f0 is None else f0
+    evals = (
+        list(map(fun, shifts[:num_frequency])) + [f0] + list(map(fun, shifts[num_frequency + 1 :]))
+    )
+    evals = anp.asarray(evals, like=interface)
+
+    x0 = anp.asarray(np.float64(0.0), like=interface) if x0 is None else x0
+
+    def _reconstruction(x):
+        """Univariate reconstruction based on equidistant shifts and Dirichlet kernels."""
+        _x = x - x0 - shifts
+        return qml.math.tensordot(
+            qml.math.sinc(a * _x) / qml.math.sinc(b * _x),
+            evals,
+            axes=[[0], [0]],
+        )
+
+    return _reconstruction
+
+
+_warn_text_f0_ignored = (
+    "The provided value of the function at zero will be ignored due to the "
+    "provided shift values. This may lead to additional evaluations of the "
+    "function to be reconstructed."
+)
+
+
+def _reconstruct_gen(fun, spectrum, shifts=None, x0=None, f0=None, interface=None):
+    r"""Reconstruct a univariate (real-valued) Fourier series with given spectrum.
+
+    Args:
+        fun (callable): Univariate finite Fourier series to reconstruct.
+            It must have signature ``float -> float`` .
+        spectrum (Collection): Frequency spectrum of the Fourier series;
+            non-positive frequencies are ignored.
+        shifts (Sequence): Shift angles at which to evaluate ``fun`` for the reconstruction
+            Chosen equidistantly within the interval :math:`[0, 2\pi/f_\text{max}]`
+            if ``shifts=None`` , where :math:`f_\text{max}` is the biggest
+            frequency in ``spectrum``.
+        x0 (float): Center to which to shift the reconstruction.
+            The points at which ``fun`` is evaluated are *not* affected
+            by ``x0`` .
+        f0 (float): Value of ``fun`` at zero; If :math:`0` is among the ``shifts``
+            and ``f0`` is provided, one evaluation of ``fun`` is saved.
+        interface (str): Which auto-differentiation framework to use as
+            interface. This determines in which interface the output
+            reconstructed function is intended to be used.
+
+    Returns:
+        callable: Reconstructed Fourier series with :math:`R` frequencies in ``spectrum`` .
+        This function is a purely classical function. Furthermore, it is fully differentiable.
+    """
```

Comment:
even though this is a private function, it's really nice that you have put the effort in to write such a detailed docstring --- will be a great help to future developers!